### ISSUE TITLE
Starting/stopping nodes should allow better automation

### ISSUE THREAD

**Issue by vstax** [leo-project/leofs#790]
There are 3 tickets related to each other: this one, #791 and #792. I'm not sure about solutions for all these problems yet, first I'd like to just outline them as I see it (which may be one-sided); if we can agree on the problems and possible solutions, I'm willing to implement them (as much as I can do it, at least). I split them into different tickets because any of them might become too long and it would be too annoying if they're together. Anyhow, about better options for starting/stopping nodes.

**Use case 1**: I'm installing software upgrades and want to reboot storage nodes one by one. Desired steps: if interrupting requests from applications isn't a problem - just execute "reboot" on nodes one by one, checking manager node directly or through monitoring to make sure last node goes into "running" state before rebooting next one. If interrupting shouldn't happen - execute "suspend <node>", then reboot that node, when it's available, execute "resume <node>" and repeat with the next one.

**Use case 2**: I need to reboot development cluster, so I just reboot all VMs and want it to correctly go down & up (I don't care what happens to other applications, or maybe I stopped them before - I just want LeoFS itself to shutdown & startup corectly).

**If it's just about the storage nodes (use case 1)**. Current scripts, e.g. "leo_storage" mimic SysV init scripts but don't have description/priorities in `chkconfig` format to be directly used as such, also copying it or executing from different path will break its logic on picking RUNNER_SCRIPT_DIR. It also doesn't install to init.d (obviously, since it's unknown what exactly node will be running on this system) and won't update if a copy is made. There are two ways of solving this issue:
1. Modifying startup script to fully confirm with SysV and adding ability to work from symlink or copy in different place, and actually creating that symlink or copy during installation
2. Wrapping it in different, simple and fully complaint init script (systemd unit file for systemd-supported systems), and installing these wrappers

Second solution might be easier given that one doesn't need to solve the problem of working from different directory, it can be worked around (but: there might be need to solve it anyway) and it's possible to provide more agile / easier to manage and modify systemd script (but: given that it'll only run shell wrapper which will run actual application which will fork, best ways to control the service from systemd won't work. Maybe it's still be better since systemd will create cgroup and can track processes through it, but I need to learn more about that to be sure if it gives any advantages). Completely modifying startup script, splitting it to few minor ones and getting rid of fork is possible as well, but annoying because development systems and non-systemd systems will have to work with old one, so two will need to be supported. Anyhow, at least some ways are possible.

**If it's about other nodes as well** (the case where I want to restart VMs holding development cluster, for example), there are additional problems:

1. Gateways status doesn't switch to "running" until at least one storage nodes is running (however, I don't think it'll abort or something, it seems to just wait patiently until that, so this isn't a real problem)
2. If storage nodes are starting before manager nodes, they will silently fail and quit. However, they do wait for some timeout before that (wonder if it can be tweaked?). Maybe trying to restart them a few times as they fail is a good idea, not sure. This kind of unreliable behaviour (when some storage nodes might fail to launch in one case and start fine in the other case) is annoying, but probably not a show stopper.
3. Starting both managers at once, or starting slave before starting manager can cause problems - at least until #562, #582 and (maybe?) #559 are fixed. Problems ranging from failing to launch once to even possibility of broken database, with need to manual recovery. Is it true or not? If it is, then auto-starting managers (and thus whole cluster) is impossible until these bugs are fixed. Or there can be some simple workaround to specifically avoid these problems?

If no workarounds for manager / full cluster are possible right now, then at least automatic start and restart of storage (can also be used for gateways) nodes on reboot can be pursued.

**There is big problem related to upgrades as well**. The init script needs to find the directory with binaries of current version (e.g. you start /etc/init.d/leo_storage, but it needs to find /usr/local/leofs/1.3.4/leo_storage). In Debian, only one version can be installed, however old directories might still remain. On EL, multiple versions can be installed, and it's a good question whether startup script should always go for the lastest version or let the user pick version.

For simplicity sake, let's assume that we always want script to handle the latest version for now. This can lead to some problems like: when not using global config files, user tries to stop the node right after new version is installed, but config files aren't changed for that new version. It will likely fail because old version is assuming different config files compared to new ones.

Alternatives can be used for tracking this latest version; e.g. `/usr/local/leofs/latest -> /etc/alternatives/leofs -> /usr/local/leofs/1.3.4`; however, correct implementation of this can be surprisingly nontrivial. Alternatives manage the version in either manual mode (any time you want it to change, you "force" it to point at the version you want), or with automatic versioning that depends on priority used for each alternative - highest one wins. For Debian, manual mode will work - the link will always point to currently installed version. It's more complicated for EL, where installing an older version in manual mode will change that link incorrectly. It can either skip installation of alternative if newer version is installed (which is obscure in implementation and will cause problems if newer version is removed while keeping the older one) or somehow convert current version to ever-growing integer and use it as priority. I've tried to use number of git commits as priority but failed because I just can't see a way to make rpm use the number I got at %build stage in %post script (i.e. make %post more dynamic), maybe there is some trick but all my tries have failed. Which probably just shows that it's a wrong path to take.
For the record, our current solution involves manual alternative, used as `/usr/local/leofs/current` symlink, but it's managed by chef cookbook, not rpm scripts. So after new version is installed on all nodes, one changes ['leofs']['version'] attribute and after that alternative is changed if the version is different when Chef Client is executed. So that completely separates rpm installation / upgrade from actual version upgrade, which might be kind of nice but won't work for anyone who didn't set up their environment for such things.

To summarize, the options I can see (for EL systems) are:
1. Forbid installation of multiple version of leofs package, then use manual alternatives mode - this will work, but is non-compatible solution, which is probably not an option (though, with global config feature - #791 - changing versions becomes as easy as installing different rpm..)
2. Use --auto with same priority (just like it's done currently for leofs-adm alternative). When installing new version, the script will point to the older version until old version is removed. I don't know if this is acceptable or not... The upgrade instructions then will be kind of "install new version, stop node (will use older version), check config files, copy if needed to new version, *remove* older version, then start node (will use new version)". If someone tries to start node before removing old version, it'll start on old version. Which is annoying.. but maybe not the end of the world and small price to pay for allowing multiple installations? Not sure.
3. Do not use alternatives here at all. Instead, package installation script (init.d/systemd/upstart) and hardcode version in there. Update this script on every upgrade. However, because two different versions of script can't be installed at the same place (same problem which forced `/usr/local/bin/leofs-adm` to be changed from real file to alternatives), the script will have to be copied over at installation time, which is problem with determining which version is to be used all over again. Also, it would make "multiple nodes of the same type" feature (#792) possible to implement only for systemd systems, not upstart/SysV.

**Comment by mocchira** [comment]
Thanks for your proposal. The overall problems you want to solve seems valid to me so let's move on how to implement those.

> Second solution might be easier given that one doesn't need to solve the problem of working from different directory, it can be worked around (but: there might be need to solve it anyway) and it's possible to provide more agile / easier to manage and modify systemd script (but: given that it'll only run shell wrapper which will run actual application which will fork, best ways to control the service from systemd won't work. Maybe it's still be better since systemd will create cgroup and can track processes through it, but I need to learn more about that to be sure if it gives any advantages). Completely modifying startup script, splitting it to few minor ones and getting rid of fork is possible as well, but annoying because development systems and non-systemd systems will have to work with old one, so two will need to be supported. Anyhow, at least some ways are possible.

Second solution looks good to me as Pros outweigh Cons (best ways to control the service from systemd won't work) and Cons might be solvable. I will wait for your survey about systemd.

> If storage nodes are starting before manager nodes, they will silently fail and quit. However, they do wait for some timeout before that (wonder if it can be tweaked?). Maybe trying to restart them a few times as they fail is a good idea, not sure. This kind of unreliable behaviour (when some storage nodes might fail to launch in one case and start fine in the other case) is annoying, but probably not a show stopper.
> (wonder if it can be tweaked?

Cannot. 

> Starting both managers at once, or starting slave before starting manager can cause problems - at least until #562, #582 and (maybe?) #559 are fixed. Problems ranging from failing to launch once to even possibility of broken database, with need to manual recovery. Is it true or not? If it is, then auto-starting managers (and thus whole cluster) is impossible until these bugs are fixed. Or there can be some simple workaround to specifically avoid these problems?

The possibility you face the problem caused by #582 would be very low while the one caused by #562 would be relatively high, yes you need manual recovery in either case. however the problem caused by #562 can be dealt with by wiping out mnesia directories/files and restarting storage again, It might be possible to provide auto-starting managers without fixing bugs while the uncertainty due to #582 exists.

> Forbid installation of multiple version of leofs package, then use manual alternatives mode - this will work, but is non-compatible solution, which is probably not an option (though, with global config feature - #791 - changing versions becomes as easy as installing different rpm..)

I'd prefer this one along with #791 (I agreed with the global config idea), though as you said, it's non-compatible solution so at least we need to hear the voice from existing users and decide based on the feedback.

**Comment by vstax** [comment]
**Regarding systemd process tracking**. The best way seems to be using "notify" type of service, which sends a ping to systemd at the end of its initialization to indicate that service is up & running. For example, I think that storage node is supposed to do this if it actually manages to connect to manager node and plans to go on with serving requests. For gateway - probably when it finishes starting up, even if it's not ready to serve because no storage nodes are available? For manager, I have no idea (there is state when manager responds to "status" command, but if for example slave manager isn't running, freshly started master won't respond to that command as well; obviously this state still needs to be treated as "finished initializing". I'm just speculating here, I'm not really sure about exact places for each type of node.

With this, systemd correctly tracks "starting up" -> "started", "starting up" -> "failed to start and exited", "starting up" -> "started" -> "crashed" states which produces more informative logs, simplifies error handling, allows to create dependencies and run commands that are executed after the service actually has started and so on. (this also means that on single-node setup, you can technically make master start up first, slave only after that, and other nodes only after that. Won't work on real cluster with single node per system, though)

While other types of services exist (like simple and forking), there will be certain kinds of problems with them for something as complicated as LeoFS node which goes through multiple internal startup phases; notify type is definitely the way to go. However, it requires application support.

Rabbitmq project shows two ways of doing it from erlang, by using `sd_notify` library, implemented like this: https://github.com/rabbitmq/rabbitmq-server/pull/52 and by executing "systemd-notify" command, implemented like this https://github.com/rabbitmq/rabbitmq-server/pull/573 and https://github.com/rabbitmq/rabbitmq-server/pull/574 (there is https://github.com/rabbitmq/rabbitmq-server/issues/664 describing the reasons for these changes). So the final version they use just tries to execute "systemd-notify --ready", always and without any error handling (`os:cmd("systemd-notify --ready"),`) - and that's about it.

Assuming LeoFS nodes execute this after finishing initialization, the changes to current startup script include:
1. Splitting off gen_config() function to separate script or adding specialized "gen_config" command which only calls gen_config(); might also fix a bug in process, currently when config generation fails because of being unable to write new app.config file or other problem, node still tries to start with old app.config. If config generation is separated, it will be called both from current script and systemd service
2. Greatly simplified code from "start" and "stop" goes to systemd service - basically just `run_erl` in foreground for start and `nodetool stop` for stop (code in current script remains unchanged). This has a downside that code that exports variables duplicates between current script and systemd service, but I think it's not a huge problem. Can add a comment about that to both as a reminder, too.
3. I'm not sure about current "restart" and "reload", what's the difference between them, usage-wise? By definition systemd never does "restart" - it will go through stop, then start for restart. It can support "reload" command for re-reading config files (usually implemented as sending SIGHUP) - but I think neither of these are equivalent. I think there is no point in assigning this "nodetool restart" for "reload", no?

Caveat: trying to start service with current script then stop through systemd service, or vice versa will fail. Not that this is a problem per se, but documentation needs to be tweaked so that no one misunderstands and tries to follow it exactly when it says `/usr/local/leofs/<version>/leo_storage/bin/leo_storage stop`, for example even when using new services.

Another idea I had is to split environment variables initialization to separate file from main script. It might make overall solution cleaner, need to experiment with that.

Non-systemd service: I think upstart service will do; it looks cleaner than SysV-style wrapper. The recommended way there is to make application run in foreground; however, it doesn't have systemd features to correctly track startup phase / finish of startup, thus it'll only be able to start and stop, no real status tracking (other than process being alive) or dependencies on that. I think since it tracks process the stop phase can be just "nodetool stop", without current wrapper which checks if process is still alive, but I need to experiment with that to be sure.

@mocchira - can you please help with pointing where exactly "systemd-notify" call can be placed for each type of node (at least for the first experiments)?

**Regarding stable symlink/alternative to the currently installed version for EL**. Yes, scrapping multi-version install might be for the best; no matter how fun current feature to have multiple versions installed at once, it creates tons of problems when trying to do something complicated as this. It's forbidden on Debian for a reason; while RPM allows it, limitations are quite severe (all shared files must be precisely the same, including contents and attributes). I can remember just two packages that can be installed in parallel on RHEL and Fedora systems - java and kernel; java employs very complicated ways of managing alternatives system for that and I've seen it to break quite a few times. Kernel does this by splitting all the shared parts to different packages, for which only one version can be present and lots of scripts/trickery as well. With the use of wrappers and other scripts people employ installing multiple versions of python, ruby, some other tricks but it's done so they can exist and used in parallel, no one tries use that kind of system for upgrading while keeping the same paths to config files.

So if a reliable solution which won't break under various conditions is required, it's about either scrapping multi-version installations, or completely different packaging, where scripts, config files and data directories and so on are split off to package that exists only in single version and must be upgraded, while just the subpackage containing binaries / runtime can be installed in multiple versions - kind of like kernel does it. This is possible, but quite complicated in packaging; upgrading becomes more complicated as well, e.g. you can't exactly do upgrade with single "rpm" command anymore, it'll be "rpm -U" for one package and "rpm -i" for another, since unlike yum/dnf, rpm doesn't allow to specify complicated transactions from command line. But that means much more incompatible change overall compared to forbidding multi-version installation while keeping structure of current package.

**Comment by mocchira** [comment]
@vstax Thanks for the survey and further digging.

> @mocchira - can you please help with pointing where exactly "systemd-notify" call can be placed for each type of node (at least for the first experiments)?

Place "systemd-notify" before the line I denoted at the following.

- manager(both master/slave): https://github.com/leo-project/leofs/blob/1.3.5-rc4/apps/leo_manager/src/leo_manager_sup.erl#L401

- storage: https://github.com/leo-project/leofs/blob/1.3.5-rc4/apps/leo_storage/src/leo_storage_app.erl#L223

- gateway: 
  - https://github.com/leo-project/leofs/blob/1.3.5-rc4/apps/leo_gateway/src/leo_gateway_app.erl#L409 (started but storage(s) are NOT running)
  - https://github.com/leo-project/leofs/blob/1.3.5-rc4/apps/leo_gateway/src/leo_gateway_app.erl#L465 (started and also storage(s) are running)

The below are answers to your questions in the previous comment (please let me know if I'm missing something).

> I'm not sure about current "restart" and "reload", what's the difference between them, usage-wise? 

It's too erlang specific thing and only meaningful for those who understand how the every piece of code are constructed so please don't mind about it.

> It can support "reload" command for re-reading config files (usually implemented as sending SIGHUP) - but I think neither of these are equivalent. I think there is no point in assigning this "nodetool restart" for "reload", no?

Right. there is no equivalent for "reload" in the current startup script however in future, we may add "reload" argument for https://github.com/leo-project/leofs/issues/673.

> Another idea I had is to split environment variables initialization to separate file from main script. It might make overall solution cleaner, need to experiment with that.
> Non-systemd service: I think upstart service will do; it looks cleaner than SysV-style wrapper. The recommended way there is to make application run in foreground; however, it doesn't have systemd features to correctly track startup phase / finish of startup, thus it'll only be able to start and stop, no real status tracking (other than process being alive) or dependencies on that. I think since it tracks process the stop phase can be just "nodetool stop", without current wrapper which checks if process is still alive, but I need to experiment with that to be sure.

Sounds good to me.

> Regarding stable symlink/alternative to the currently installed version for EL.

Thanks for the insightful comment that give me the confidence that scrapping multi-version install might be for the best.

**Comment by vstax** [comment]
@mocchira Thank you, I will check how it works with `systemd-notify` at the places you mentioned.

Please take a look at https://github.com/leo-project/leofs_package/pull/6

It would be really nice if it's possible to release 1.3.5 with this change, because with this old config files (current style) aren't removed during upgrade if they're changed. It allows to directly upgrade rpm and not lose old changes to config files. If multi-version install is disabled in some future version, then 1.3.5 becomes a version starting from which you can do upgrades directly without losing changes to config files - all that has to be done is asking them to upgrade to 1.3.5 first. I did similar change to deb package as well but it works so poorly, leaving all the config files after each upgrade so I had to scrap it.

Of course, users that copy config files from some other place and not from directory with previous version won't be affected by this change (and on Debian/Ubuntu it's the only way to do it right now anyway), but it's better to have it in case current users of rpm versions had expectations that they can copy config files from old version into new version.

**Comment by mocchira** [comment]
@vstax 

> I will check how it works with systemd-notify at the places you mentioned

Let me know if you have some trouble.

> Please take a look at leo-project/leofs_package#6

Just in case (I already commented on PR), LGTM and I think that can be merged before 1.3.5 released. Please wait for a while as others still can't spare time for these changes.

**Comment by vstax** [comment]
@mocchira Great, thanks for looking at that PR.

Regarding this issue. Still experimenting with how to smooth things out with the external service but I'd like to note a few things that I discovered so far (mostly unfortunate). You don't need to do anything yet but when you have some free time, I'd like to have your opinion on "how to execute" matter and heartbeat test. Not sharing full code yet because it's really WIP, just some pieces to get idea on how it might look in the end.

There are _lots_ of environmental variables used one way or another in current script, including ones that are extracted in complicated way from generated config files. Redefining them all is bad idea, so I added a command which generates config files and them dumps all the required environments to temporary file; systemd can execute a command and read variables that it creates. Kind of like this (through this one has some extras I used while experimenting):
```
+    gen_environment)
+	shift  # for correct CONSOLE_COMMAND
+        cat << EOT > /run/leofs/${SCRIPT}.env
+# NODETOOL="$NODETOOL"
+# NAME_ARG=$NAME_ARG
+# COOKIE_ARG=$COOKIE_ARG
+
+ERTS_PATH=$ERTS_PATH
+PIPE_DIR=$PIPE_DIR
+RUNNER_LOG_DIR=$RUNNER_LOG_DIR
+
+ROOTDIR=$RUNNER_BASE_DIR
+BINDIR=$ERTS_PATH
+EMU=beam
+PROGNAME=$SCRIPT
+
+APP_VSN=$APP_VSN
+BOOTFILE=$SCRIPT
+ERLEXEC_MODE=$ERLEXEC_MODE
+CONFIG_PATH=$CONFIG_PATH
+VMARGS_PATH=$VMARGS_PATH
+
+# CONSOLE_COMMAND="exec $ERTS_PATH/erlexec -noinput -boot $RUNNER_BASE_DIR/releases/$APP_VSN/$BOOTFILE -mode $ERLEXEC_MODE -config $CONFIG_PATH -args_file $VMARGS_PATH -- console"
+# HEART_COMMAND="$BINDIR/run_erl $PIPE_DIR $RUNNER_LOG_DIR $CONSOLE_COMMAND"
+# HEART_COMMAND="$ERTS_PATH/erlexec -noinput -boot $RUNNER_BASE_DIR/releases/$APP_VSN/$BOOTFILE -mode $ERLEXEC_MODE -config $CONFIG_PATH -args_file $VMARGS_PATH -- console"
+EOT
+        ;;
```

**"systemd-notify --ready" is not working under non-root user**. Well it does sometimes, but mostly doesn't, I'm running into bug https://lists.freedesktop.org/archives/systemd-devel/2014-April/018797.html - basically it exists right away and by the time systemd looks at what cgroup the process does belong it doesn't exist anymore. Workarounds are running as root (not an option), passing PID of main process to this command so it reports that pid to systemd and not its own (not sure; need to check why rabbitmq decided to stop doing that, maybe it's not reliable enough as well) and actually doing call with sd-notify library from main process (it's adopted by couchdb, for example; I still fail to understand why rabbitmq decided to it otherwise). I gave up on this part for now and working with "simple" service, this can be changed later.

**The fact that Erlang starts `epmd` isn't working well with systemd**, because when service stops it will kill everything that process has spawned, including epmd.. and epmd is supposed to stay alive if there are any other processes left. It needs to be split off to separate service (e.g. https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=709754). On a positive side, if Erlang 17 or higher is built with systemd support, epmd supports socket activation by systemd and it's possible to write socket-type service that spawns epmd only when the need arises. Still, it will work even with permanently running epmd (probably the biggest question is - under which user to spawn it? "leofs" or "root"? I suppose "leofs" is fine, if someone needs to run some other Erlang software on the same system as LeoFS it's up to that person to tweak these settings). Anyhow this shouldn't be a serious problem, just something to keep in mind.

**Systemd can accept parameters from environment variables**, but path to binary must be absolute; I can see the reasons for that, but scripts needs to be created during build by filling a few macros because of that. Example service which I'm experimenting with now is kind of like this:
```
[Service]
Type=simple
NotifyAccess=all
User=leofs

ExecStartPre=/usr/local/leofs/current/leo_manager_1/bin/leo_manager gen_environment
EnvironmentFile=/run/leofs/leo_manager.env

Environment="CONSOLE_COMMAND=exec $ERTS_PATH/erlexec -noinput -boot /usr/local/leofs/current/leo_manager_1/releases/${APP_VSN}/${BOOTFILE} -mode $ERLEXEC_MODE -config $CONFIG_PATH -args_file $VMARGS_PATH -- console"
#Environment="HEART_COMMAND=/usr/local/leofs/current/leo_manager_1/erts-8.3/bin/run_erl $PIPE_DIR $RUNNER_LOG_DIR $CONSOLE_COMMAND"
Environment="HEART_COMMAND=/usr/local/leofs/current/leo_manager_1/erts-8.3/bin/erlexec -noinput -boot /usr/local/leofs/current/leo_manager_1/releases/${APP_VSN}/${BOOTFILE} -mode $ERLEXEC_MODE -config $CONFIG_PATH -args_file $VMARGS_PATH -- console"

StandardOutput=syslog+console
StandardError=syslog+console

WorkingDirectory=/usr/local/leofs/current/leo_manager_1

#ExecStart=/usr/local/leofs/current/leo_manager_1/erts-8.3/bin/run_erl $PIPE_DIR $RUNNER_LOG_DIR "$CONSOLE_COMMAND"
ExecStart=/usr/local/leofs/current/leo_manager_1/erts-8.3/bin/erlexec -noinput -boot /usr/local/leofs/current/leo_manager_1/releases/${APP_VSN}/${BOOTFILE} -mode $ERLEXEC_MODE -config $CONFIG_PATH -args_file $VMARGS_PATH -- console

TimeoutStopSec=60
ExecStop=/usr/local/leofs/current/leo_manager_1/bin/leo_manager stop
```

I replaced all paths to "erts-8.3" to absolute here, even in arguments because it doesn't make sense to keep them like that.


**Another question I'm thinking about: how to launch.** The original script has quite complicated way of executing `to_erl` which executes the same script again with `console` argument which executes `erlexec`.. and heartbeat is set to the same script again. Some ideas that I found:
1. rabbitmq launches through the same script regardless of systemd, but systemd passes argument which makes it launch the real process in foreground. This is neat, but current launch scripts are way too complicated for this to make sense
2. Some people are saying not to use `to_erl` with systemd (because it captures stdout/stderr to logs by itself anyway) and use `relx`-generated executable instead (which is not an option here). Sure, I can make the things work by running both `erlexec` without `to_erl` wrapper and with it for now. There will be no logs that `to_erl` creates but it all exists in journald instead. Not 100% sure if it's worth to drop `to_erl` or not, main reasons for dropping it is that it doesn't add much under systemd and may hide errors from `erlexec` to its own log, while without it you can see them directly in journald. Plus, well, you can see in example above how both solutions look in service file... I don't even want to comment that.

Anyhow, "-daemon" argument to erlexec is dropped and "-noinput" is required because systemd obviosly can't provide stdin.

**Hearbeat**: I wish I knew how to test it properly. Well, I can execute original script with "restart" and it works fine. Btw, I noticed that original script with "reboot" maybe is broken? It closes the log files when executed but the process is kind of dead after that. Even when service is launched through original script. Any suggestions on some command that I can do so that original process re-executes itself through "HEARBEAT_COMMAND"?

**Stopping**: well, this is another disappointment, systemd expects "stop" operation to be synchronous. After the "ExecStop" command returns, it assumes that nothing is supposed to be living anymore so it sends signal to anything that was left behind. Sure, it sends nice signal first and KILL only after timeout but still, "nodeping stop" just doesn't work here. I'm calling original script for now as it has code to wait for process.

**Comment by mocchira** [comment]
@vstax Thanks for surveying on systemd. I almost got what (unfortunate things) you shared on the prev comment.

> Redefining them all is bad idea, so I added a command which generates config files and them dumps all the required environments to temporary file

Looks good to me.

> (probably the biggest question is - under which user to spawn it? "leofs" or "root"? I suppose "leofs" is fine, if someone needs to run some other Erlang software on the same system as LeoFS it's up to that person to tweak these settings). Anyhow this shouldn't be a serious problem, just something to keep in mind.

leofs is fine and agree that's not a serious problem.

> Hearbeat: I wish I knew how to test it properly. Well, I can execute original script with "restart" and it works fine. Btw, I noticed that original script with "reboot" maybe is broken? It closes the log files when executed but the process is kind of dead after that. Even when service is launched through original script. Any suggestions on some operation that I can do so that original process re-executes itself through "HEARBEAT_COMMAND" (like, idk, attach remote console - which works - and execute something which will freeze the process and makes heartbeat kick in)?

Thanks for pointing that out! reboot is broken as of now, I'll send PR later and regarding re-executes itself through HEART_COMMAND, yes "reboot" is what you want so please check it out once the PR get merged.

> Some people are saying not to use to_erl with systemd (because it captures stdout/stderr to logs by itself anyway) and use relx-generated executable instead (which is not an option here). Sure, I can make the things work by running both erlexec without to_erl wrapper and with it for now. There will be no logs that to_erl creates but it all exists in journald instead. Not 100% sure if it's worth to drop to_erl or not, main reasons for dropping it is that it doesn't add much under systemd and may hide errors from erlexec to its own log, while without it you can see them directly in journald. Plus, well, you can see in example above how both solutions look in service file... I don't even want to comment that.

(Just in case, let me confirm that **to_erl** you mentioned at the prev comment isn't **run_erl**?)
Anyhow, Cons I come up with to drop **run_erl** is "attach" command will no longer work anymore. however "remote_console" should keep working and it can be substituted so IMHO it's acceptable. @yosukehara what do you think about it?

**Comment by mocchira** [comment]
@vstax reboot is working now as expected, however there may be one problem. With getting reboot worked, all erlang processes get to be monitored by **heart** command which spawns an erlang process and restart the process whenever that process goes down. let me know if this behavior would bother you.

As for dropping **run_erl**, we reached the consensus we could drop it.

**Comment by vstax** [comment]
@mocchira Yes, of course I meant `run_erl`, sorry for the confusion! And yes, "remote_console" works.

The fact that it spawns another process could be somewhat a problem, as for non-forking service systemd expect main process to act as supervisor which may start and restart its children but does not quit itself as long as it's working (it tracks whether service is alive through this process, and can send shutdown signal to this process only, expecting it to cleanup its children in a safe way). I wish it would just exec the new process instead without forking. That said, I'll experiment with it to check how it works, so it's too early to make any changes (plus, after systemd I'll check how upstart works, which might change things as well).

**Comment by vstax** [comment]
### Heartbeat

Just like I thought, Erlang heartbeat won't work with systemd, restart of main process is unacceptable for non-forking daemon:

```
[root@leo-m1 ~]# systemctl status leofs-manager-slave
● leofs-manager-slave.service
   Loaded: loaded (/etc/systemd/system/leofs-manager-slave.service; static; vendor preset: disabled)
   Active: active (running) since Чт 2017-08-03 17:57:37 MSK; 29s ago
  Process: 3232 ExecStop=/usr/local/leofs/current/leo_manager_1/bin/leo_manager stop (code=exited, status=1/FAILURE)
  Process: 3407 ExecStartPre=/usr/local/leofs/current/leo_manager_1/bin/leo_manager gen_environment (code=exited, status=0/SUCCESS)
 Main PID: 3498 (beam.smp)
   CGroup: /system.slice/leofs-manager-slave.service
           ├─3498 /usr/local/leofs/current/leo_manager_1/erts-8.3/bin/beam.smp -A 32 -K true -P 1048576 -- -root /usr/local...
           ├─3537 erl_child_setup 1024
           ├─3547 heart -pid 3498
           ├─3635 inet_gethost 4
           └─3636 inet_gethost 4

[root@leo-m1 ~]# /usr/local/leofs/current/leo_manager_1/bin/leo_manager reboot
[root@leo-m1 ~]# systemctl status -l leofs-manager-slave
● leofs-manager-slave.service
   Loaded: loaded (/etc/systemd/system/leofs-manager-slave.service; static; vendor preset: disabled)
   Active: failed (Result: exit-code) since Чт 2017-08-03 17:58:21 MSK; 34s ago
  Process: 3753 ExecStop=/usr/local/leofs/current/leo_manager_1/bin/leo_manager stop (code=exited, status=1/FAILURE)
  Process: 3498 ExecStart=/usr/local/leofs/current/leo_manager_1/erts-8.3/bin/erlexec -noinput -heart -boot /usr/local/leofs/current/leo_manager_1/releases/${APP_VSN}/${BOOTFILE} -mode $ERLEXEC_MODE -config $CONFIG_PATH -args_file $VMARGS_PATH -- console (code=killed, signal=KILL)
  Process: 3407 ExecStartPre=/usr/local/leofs/current/leo_manager_1/bin/leo_manager gen_environment (code=exited, status=0/SUCCESS)
 Main PID: 3498 (code=killed, signal=KILL)

авг 03 17:58:18 leo-m1.dev.cloud.lan erlexec[3498]: * closing log file is /var/log/leofs/leo_manager_1/app/error.20170803.17.1
авг 03 17:58:18 leo-m1.dev.cloud.lan erlexec[3498]: * closing log file is /var/log/leofs/leo_manager_1/app/info.20170803.17.3
авг 03 17:58:18 leo-m1.dev.cloud.lan erlexec[3498]: * closing log file is /var/log/leofs/leo_manager_1/app/cmd_history.20170803.17.1
авг 03 17:58:18 leo-m1.dev.cloud.lan erlexec[3498]: * closing log file is /var/log/leofs/leo_manager_1/app/error.20170803.17.1
авг 03 17:58:18 leo-m1.dev.cloud.lan erlexec[3498]: * closing log file is /var/log/leofs/leo_manager_1/app/info.20170803.17.3
авг 03 17:58:21 leo-m1.dev.cloud.lan systemd[1]: leofs-manager-slave.service: main process exited, code=killed, status=9/KILL
авг 03 17:58:21 leo-m1.dev.cloud.lan leo_manager[3753]: Node 'manager_1@192.168.3.51' not responding to pings.
авг 03 17:58:21 leo-m1.dev.cloud.lan systemd[1]: leofs-manager-slave.service: control process exited, code=exited status=1
авг 03 17:58:21 leo-m1.dev.cloud.lan systemd[1]: Unit leofs-manager-slave.service entered failed state.
авг 03 17:58:21 leo-m1.dev.cloud.lan systemd[1]: leofs-manager-slave.service failed.
```

Typically, suggestions are "implementation of Erlang's heart isn't compatible with systemd, do not use heart, use systemd own process monitoring and restart functionality". Apparently there also is advantage in case process is killed by OOM - heart won't work in that case, but systemd's restart will (for that reason, some suggest to use service wrapper with external heartbeat, e.g. `monit`, even on non-systemd service). Anyhow, systemd restart solution will look something like this: https://github.com/klacke/yaws/blob/master/scripts/systemd/yaws.service - funny enough, it kind of works even with current "reboot" command - systemd notices that old process died, kills every other process in cgroup and then restarts original daemon 5 seconds later:
```
авг 03 18:11:37 leo-m1.dev.cloud.lan erlexec[3997]: * closing log file is /var/log/leofs/leo_manager_1/app/cmd_history.20170803.18.1
авг 03 18:11:37 leo-m1.dev.cloud.lan erlexec[3997]: * closing log file is /var/log/leofs/leo_manager_1/app/error.20170803.18.1
авг 03 18:11:37 leo-m1.dev.cloud.lan erlexec[3997]: * closing log file is /var/log/leofs/leo_manager_1/app/info.20170803.18.1
авг 03 18:11:37 leo-m1.dev.cloud.lan erlexec[3997]: * closing log file is /var/log/leofs/leo_manager_1/app/cmd_history.20170803.18.1
авг 03 18:11:37 leo-m1.dev.cloud.lan erlexec[3997]: * closing log file is /var/log/leofs/leo_manager_1/app/error.20170803.18.1
авг 03 18:11:37 leo-m1.dev.cloud.lan erlexec[3997]: * closing log file is /var/log/leofs/leo_manager_1/app/info.20170803.18.1
авг 03 18:11:40 leo-m1.dev.cloud.lan erlexec[3997]: heart: Thu Aug  3 18:11:40 2017: Erlang has closed.
авг 03 18:11:40 leo-m1.dev.cloud.lan systemd[1]: leofs-manager-slave.service: main process exited, code=killed, status=9/KILL
авг 03 18:11:40 leo-m1.dev.cloud.lan leo_manager[4192]: Node 'manager_1@192.168.3.51' not responding to pings.
авг 03 18:11:40 leo-m1.dev.cloud.lan systemd[1]: leofs-manager-slave.service: control process exited, code=exited status=1
авг 03 18:11:40 leo-m1.dev.cloud.lan systemd[1]: Unit leofs-manager-slave.service entered failed state.
авг 03 18:11:40 leo-m1.dev.cloud.lan systemd[1]: leofs-manager-slave.service failed.
авг 03 18:11:45 leo-m1.dev.cloud.lan systemd[1]: leofs-manager-slave.service holdoff time over, scheduling restart.
авг 03 18:11:45 leo-m1.dev.cloud.lan systemd[1]: Starting leofs-manager-slave.service...
авг 03 18:11:46 leo-m1.dev.cloud.lan systemd[1]: Started leofs-manager-slave.service.
авг 03 18:11:46 leo-m1.dev.cloud.lan erlexec[4342]: heart_beat_kill_pid = 4342
авг 03 18:11:46 leo-m1.dev.cloud.lan erlexec[4342]: id:leo_logger_file_i, path:"/var/log/leofs/leo_manager_1/app", filename:"info"
авг 03 18:11:46 leo-m1.dev.cloud.lan erlexec[4342]: * opening log file is /var/log/leofs/leo_manager_1/app/info.20170803.18.2
авг 03 18:11:46 leo-m1.dev.cloud.lan erlexec[4342]: id:leo_logger_file_e, path:"/var/log/leofs/leo_manager_1/app", filename:"error"
авг 03 18:11:46 leo-m1.dev.cloud.lan erlexec[4342]: * opening log file is /var/log/leofs/leo_manager_1/app/error.20170803.18.1
авг 03 18:11:46 leo-m1.dev.cloud.lan erlexec[4342]: id:log_id_history_log, path:"/var/log/leofs/leo_manager_1/app", filename:"cmd_history"
авг 03 18:11:46 leo-m1.dev.cloud.lan erlexec[4342]: * opening log file is /var/log/leofs/leo_manager_1/app/cmd_history.20170803.18.1
```

The same experiment but without `-heart` option:
```
авг 03 18:46:11 leo-m1.dev.cloud.lan erlexec[4848]: * closing log file is /var/log/leofs/leo_manager_1/app/cmd_history.20170803.18.1
авг 03 18:46:11 leo-m1.dev.cloud.lan erlexec[4848]: * closing log file is /var/log/leofs/leo_manager_1/app/error.20170803.18.1
авг 03 18:46:11 leo-m1.dev.cloud.lan erlexec[4848]: * closing log file is /var/log/leofs/leo_manager_1/app/info.20170803.18.4
авг 03 18:46:11 leo-m1.dev.cloud.lan erlexec[4848]: * closing log file is /var/log/leofs/leo_manager_1/app/cmd_history.20170803.18.1
авг 03 18:46:11 leo-m1.dev.cloud.lan erlexec[4848]: * closing log file is /var/log/leofs/leo_manager_1/app/error.20170803.18.1
авг 03 18:46:11 leo-m1.dev.cloud.lan erlexec[4848]: * closing log file is /var/log/leofs/leo_manager_1/app/info.20170803.18.4
авг 03 18:46:14 leo-m1.dev.cloud.lan leo_manager[5044]: Node 'manager_1@192.168.3.51' not responding to pings.
авг 03 18:46:14 leo-m1.dev.cloud.lan systemd[1]: leofs-manager-slave.service: control process exited, code=exited status=1
авг 03 18:46:14 leo-m1.dev.cloud.lan systemd[1]: Unit leofs-manager-slave.service entered failed state.
авг 03 18:46:14 leo-m1.dev.cloud.lan systemd[1]: leofs-manager-slave.service failed.
авг 03 18:46:19 leo-m1.dev.cloud.lan systemd[1]: leofs-manager-slave.service holdoff time over, scheduling restart.
авг 03 18:46:19 leo-m1.dev.cloud.lan systemd[1]: Starting leofs-manager-slave.service...
авг 03 18:46:19 leo-m1.dev.cloud.lan systemd[1]: Started leofs-manager-slave.service.
авг 03 18:46:20 leo-m1.dev.cloud.lan erlexec[5195]: id:leo_logger_file_i, path:"/var/log/leofs/leo_manager_1/app", filename:"info"
авг 03 18:46:20 leo-m1.dev.cloud.lan erlexec[5195]: * opening log file is /var/log/leofs/leo_manager_1/app/info.20170803.18.5
авг 03 18:46:20 leo-m1.dev.cloud.lan erlexec[5195]: id:leo_logger_file_e, path:"/var/log/leofs/leo_manager_1/app", filename:"error"
авг 03 18:46:20 leo-m1.dev.cloud.lan erlexec[5195]: * opening log file is /var/log/leofs/leo_manager_1/app/error.20170803.18.1
авг 03 18:46:20 leo-m1.dev.cloud.lan erlexec[5195]: id:log_id_history_log, path:"/var/log/leofs/leo_manager_1/app", filename:"cmd_history"
авг 03 18:46:20 leo-m1.dev.cloud.lan erlexec[5195]: * opening log file is /var/log/leofs/leo_manager_1/app/cmd_history.20170803.18.1
авг 03 18:46:20 leo-m1.dev.cloud.lan epmd[2414]: epmd: got partial packet only on file descriptor 6 (0)
```

A bit dubious chain of events but it doesn't seem to damage anything any more than type of restart that "reboot" command does by itself. Seems to work saner without `-heart`. I think systemd's own monitoring/restart works fine here and provides the same experience from administrator point of view.

The question here is: with `-heart` default value for `ERL_CRASH_DUMP_SECONDS` is "0", without it's "-1". Should this be changed to "0" as well when using systemd's restart manager by default (unless user explicitly sets it to -1)? It makes sense for the same reason why it's "0" if heart is enabled, generally you don't want to have tons of crash dump from auto-restarting service.


### Regarding epmd

TL;DR - I've tried various things and came to conclusion that really smooth experience is possible only for systemd-enabled epmd and socket activation (even when there is no other Erlang software), which requires Erlang to be compiled with systemd support.

In details. The state of things is like this: any LeoFS component starts epmd automatically, if it wasn't started before. If it's already running, nothing will happen, but if that epmd is stopped after that, then service which connected to it will fail in obscure way.

The problem is - epmd should never be spawned by node running under systemd, because that instance of epmd will be considered belonging to this service and will be killed whenever this node is stopped or restarted. That means that additional service is needed and any other node needs to have it as dependency, e.g.:
```
After=leofs-epmd.service
```
With this, starting service for any node will make sure that epmd service is started first, and systemd will start `leofs-epmd.service` if it's not started yet. Problem is, it's not really possible to write `leofs-epmd.service` in 100% reliable way; e.g. it contains
```
ExecStart=/usr/local/leofs/current/leo_manager_0/erts-8.3/bin/epmd
ExecStop=/usr/local/leofs/current/leo_manager_0/erts-8.3/bin/epmd -kill
```
but **if** any LeoFS node was launched on this system before, or for some other reason epmd is already running, starting this service will fail, as another instance of epmd cannot be started, thus systemd won't allow real node to start as well, because the dependency (leofs-epmd.service) isn't running. This can be solved with adding
```
ExecStartPre=-/usr/local/leofs/current/leo_manager_0/erts-8.3/bin/epmd -kill
```
to leofs-epmd.service, then it's always possible to start it, but the problem is that killing epmd like that can cause problems to any Erlang code that's running at this point. E.g. it becomes impossible to try starting leofs-epmd.service after installation of rpm/deb package and causes other obscure problems.

While it's possible to go with all this when exercising caution and making all users strictly follow documentation, I feel like it's too fragile to be used this way. The solution is socket activation (systemd feature). With it systemd always listen on epmd socket and makes it seem for any Erlang software that epmd exists, and spawns it in background when it's actually needed. The benefit here is that there are no problems with some other software starting it (other than right after the first installation of LeoFS on system where Erlang software is running right now) - no one would start it directly, and since it's fully systemd-managed, there is no need to kill it before starting, systemd always knows whether it's real epmd that is running or just systemd itself is listening on socket.

This solution looks like this, leofs-epmd.socket:
```
[Unit]
Description=Erlang Port Mapper Daemon Activation Socket

[Socket]
ListenStream=0.0.0.0:4369
Accept=false

[Install]
WantedBy=sockets.target
```
and leofs-epmd.service:
```
[Unit]
Description=Erlang Port Mapper Daemon
After=network.target
Requires=leofs-epmd.socket

[Service]
ExecStart=/usr/local/leofs/current/leo_manager_0/erts-8.3/bin/epmd -systemd
ExecStop=/usr/local/leofs/current/leo_manager_0/erts-8.3/bin/epmd -kill
Type=simple
User=leofs
Group=leofs

[Install]
Also=leofs-epmd.socket
WantedBy=multi-user.target
```

As long as `leofs-epmd.socket` is enabled (which happens on system startup), systemd will start listening on epmd socket before it's actually launched and fully manage launching it as any software demands (and no software can force-start epmd as well because socket will be in use). The only situation that needs to be covered is the first installation; but starting leofs-epmd.socket from rpm/deb installation script can almost always take care of it, and it's enough to mention it in documentation to cover the rest of cases (suggesting either to kill already-running epmd or reboot). The problem is that epmd needs to support the socket activation (`-socket` option), and for that Erlang needs to be built with systemd. Erlang 17 and higher supports it automatically if `systemd-devel` package is installed on EL (`libsystemd-dev` for Debian), the check can be forced with `configure --enable-systemd`. So while this **is** the best possible way, Erlang which is used for building official EL7 and Ubuntu 16 packages needs to be compiled with systemd support (currently it's not); obviously, one for EL6 and Ubuntu 14 cannot be compiled like that because there is no systemd there. It's easy to verify result: if Erlang compiled with systemd support, then `-socket` option appears in epmd (and shown in `epmd --help` output).

Would it be possible to arrange this?


Anyhow, except for notify-type service which properly reports its startup phase (which would be really nice to have - just look at hacks that are needed when you want to create any startup-related logic with simple service: https://bugzilla.redhat.com/show_bug.cgi?id=1112770#c0 - and it still fails to work 100% reliable: https://bugzilla.redhat.com/show_bug.cgi?id=1104193#c2) I got it all working pretty nice with systemd, I think. I gave up on "gen_environment" solution because of some minor issues, and since I'm using original script for environment and stopping anyway, I started using it for start as well. Just added an option to start `erlexec` directly and in foreground, this allowed me to remove all the environment / complex commandline mess from systemd service (since we are going through original script anyway, it can be left there, easier to maintain it all in one place anyway).

I'll look at upstart and try to create solution for it as well, however this will have to wait about 2 weeks for which I won't be able to work on anything of this at all - I'll return to these tasks after.

**Comment by mocchira** [comment]
@vstax Thanks for the detailed report.

> The question here is: with -heart default value for ERL_CRASH_DUMP_SECONDS is "0", without it's "-1". Should this be changed to "0" as well when using systemd's restart manager by default (unless user explicitly sets it to -1)? It makes sense for the same reason why it's "0" if heart is enabled, generally you don't want to have tons of crash dump from auto-restarting service.

Make sense changing to 0.

> Would it be possible to arrange this?

Yes I think possible. I'd like to discuss with other members and will provide official packages bundled with the erlang runtime compiled with systemd support.

> Anyhow, except for notify-type service which properly reports its startup phase (which would be really nice to have - just look at hacks that are needed when you want to create any startup-related logic with simple service: https://bugzilla.redhat.com/show_bug.cgi?id=1112770#c0 - and it still fails to work 100% reliable: https://bugzilla.redhat.com/show_bug.cgi?id=1104193#c2) I got it all working pretty nice with systemd, I think. I gave up on "gen_environment" solution because of some minor issues, and since I'm using original script for environment and stopping anyway, I started using it for start as well. Just added an option to start erlexec directly and in foreground, this allowed me to remove all the environment / complex commandline mess from systemd service (since we are going through original script anyway, it can be left there, easier to maintain it all in one place anyway).

Got it. Going through the original script sounds good.

> I think the options at this point are:

IMHO, 

> Ditching upstart support, no nice system-level services for Ubuntu 14 and EL6, support only systemd-based systems

seems the promising way. however there might be some non-negligible users who would expect upstart to be supported so we'd like to seek community's opinions just in case.

**Comment by mocchira** [comment]
Note:

> Yes I think possible. I'd like to discuss with other members and will provide official packages bundled with the erlang runtime compiled with systemd support.

The first official packages for 1.3.5 will be published without systemd support. we will publish the packages with systemd support later on so please wait for a while.

**Comment by mocchira** [comment]
@vstax The official packages with systemd support has been released. Let me know if you find something unexpected.

**Comment by vstax** [comment]
@mocchira 

### epmd

Erlang runtime packaged with current 1.3.5 supports systemd and works great (I checked only EL7 version so far, though), thank you.

However, the path to epmd in service file needs to be hardcoded, there are two lines like this:
```
ExecStart=/usr/local/leofs/current/leo_manager_0/erts-9.0/bin/epmd -systemd
ExecStop=/usr/local/leofs/current/leo_manager_0/erts-9.0/bin/epmd -kill
```

`leo_manager_0` part will always stay the same, but `erts-9.0` part depends on Erlang version. So there will be a macro in source file which will be expanded (probably during "make release").

Question: can it be done with reltool, or I need to write some sed snippet that does it from Makefile (not that it's too hard, just seems improper if it can be done in reltool.config)?


### About removing ability to install multiple versions on EL

If it's decided that this is the best way, I need to start creating that solution (well, after 1.3.6 release, of course) - I'm done with testing systemd service files with hardcoded paths and now need to think how to integrate all that into "make release" and package creation, and for that I need a way to ensure exact path.

It's pretty clear from e.g. epmd service (which has erts-* hardcoded) that it's nearly impossible to do a nice solution which supports multiple versions, because for example switching Erlang version (like it was done from 1.3.4 to 1.3.5) changes service file and it becomes impossible to install 1.3.4 and 1.3.5 if they both supply different service file (if they tried to supply it, that is). I think that with latest changes to package (which allows you to upgrade package first and copy the config files from old directory to new after upgrading) and also global config support (which I plan to do after this ticket, I think it's not that much work compared to this one) it should give no problems to existing users. However, about contacting / warning them beforehand, what would be the best way? A email in ML? A note on website?


### About script change to allow foreground execution

I commited (final for now, unless there will be some problems) version here: https://github.com/vstax/leofs/commit/9b799af7ced43f739d13e28dcb3c7bb11bc3e259

(I also have version where "foreground_start" is in separate section (not merged with console/console_clean); shell syntax is pretty messy so it's hard to decide if code duplication or nested if/case is the best - both look not that great)

There is a change in this script, now it always executes `nodetool chkconfig` after config generation; current version has an issue where `console` mode re-generates config, but does not check it afterwards (e.g. when restarted by heart after crash), silently running with old config if generation of new one has failed; I think it should either check config every time or not re-generate config at all.

Also, I should note that this "foreground_support" mode would probably be nice for running in docker and other similar cases.

**Comment by mocchira** [comment]
@vstax 

> Question: can it be done with reltool, or I need to write some sed snippet that does it from Makefile (not that it's too hard, just seems improper if it can be done in reltool.config)?

It's quite difficult to find the proper reltool spec as almost nothing information out there. however we've already used **{{erts_vsn}}** as the placeholder for the erlang runtime version and found out reltool.config can have **replace** instruction in the overlay section that allow us to do **sed** like operations according to https://github.com/basho/rebar/blob/master/src/rebar_reltool.erl#L411-L428.

Upon the above observations, we might be able to avoid the erts hardcoded by having
```erlang
{replace, "bin/leo_manager", "\{\{ERTS_VSN\}\}", "\{\{erts_vsn\}\}"},
```
this line into the overlay and having
```bash
## {{ERTS_VSN}} is replaced with the actual erlang runtime version by reltool
ExecStart=/usr/local/leofs/current/leo_manager_0/{{ERTS_VSN}}/bin/epmd -systemd
ExecStop=/usr/local/leofs/current/leo_manager_0/{{ERTS_VSN}}/bin/epmd -kill
```
those lines into the startup script.
(I have not confirmed the above assumption actually works yet though work I believe..)

> However, about contacting / warning them beforehand, what would be the best way? A email in ML? A note on website?

I think **A email in ML** should work as it worked (almost no complaint we've received) when we introduced the default user changing to leofs.

> About script change to allow foreground execution

All changes in https://github.com/vstax/leofs/commit/9b799af7ced43f739d13e28dcb3c7bb11bc3e259  LGTM.

**Comment by vstax** [comment]
@mocchira

Thank you, "replace" in reltool works nicely. Couldn't get that exact line to work, though - neither `\{\{ERTS_VSN\}\}` nor `{{ERTS_VSN}}` catch `{{ERTS_VSN}}`, so I switched to `%ERTS_VSN%` (maybe reltool considers {{..}} to be special and tries to expand it internally or something, but it's no big deal).

I'd be grateful if you could give me another advice regarding installation of new files so there won't be too much mess. I was thinking about temporary place for service files after "make release" (in "etc" but it can be anywhere really), which are then moved to proper place during package installation and removed from "etc". For example, there is service file `package/leo_manager_0/etc/leofs-manager-master.service` after "make release" and package generation **moves** that file to `/usr/lib/systemd/system` and it won't be present in `/usr/local/leofs/<version>/leo_manager_0/etc`.

The problem is that same `reltool.config` is the same for both managers; for main config files it's circumvented in Makefile where (just for managers) copies of separate config and schema files are made. If I put installation to reltool.config, I end up wiith this:
```
$ ls package/leo_manager_*/etc/
package/leo_manager_0/etc/:
leofs-epmd.service  leofs-manager-master.service  leo_manager.conf  leo_manager.environment
leofs-epmd.socket   leofs-manager-slave.service   leo_manager.d     leo_manager.schema

package/leo_manager_1/etc/:
leofs-epmd.service  leofs-manager-master.service  leo_manager.conf  leo_manager.environment
leofs-epmd.socket   leofs-manager-slave.service   leo_manager.d     leo_manager.schema
```

Sure, I will copy these files and remove them during package installation but I wonder if it's possible to avoid placing extra copies there in the first place. One option is to copy these service files directly in Makefile, but the code does not look good at all (and chance to do "replace" in reltool is lost). Another is not to copy these files at all (pick them directly from rpm/deb) but it has other downsides, plus what if someone wants to pick these service files without rpm/deb generation? Another is to leave like this, I suppose.
Current implementation of "make release" part is here: https://github.com/vstax/leofs/commit/99c70368dafad7405d04d6c6e608395a8c614175

Also, I'm not sure about non-Linux users - these service files will be useless for Solaris and FreeBSD, in that sense copying them to "etc" with "make release" isn't all that great as they'll have to remove them. I suppose I could move it to another make target, like "make systemd" or something, that means no "replace" with reltool but otherwise might be not a bad idea.

There are few service files so I created new directory so they're together (some related files might be placed there in the future as well - e.g. limits configuration or helper scripts). Also, since `systemctl` supports nice output and autocompletion, shortening service names is discouraged, also `_` is recommended to be replaced with `-`, so I took liberty in naming services for "leo_manager_0" and "leo_manager_1" as "leofs-manager-master" and "leofs-manager-slave".

**Comment by mocchira** [comment]
@vstax 

> Another is to leave like this, I suppose.
Current implementation of "make release" part is here: https://github.com/vstax/leofs/commit/99c70368dafad7405d04d6c6e608395a8c614175

This one looks most promising to me for now however to deal with non-linux users, we may have to follow the another option.

> Also, I'm not sure about non-Linux users - these service files will be useless for Solaris and FreeBSD, in that sense copying them to "etc" with "make release" isn't all that great as they'll have to remove them. I suppose I could move it to another make target, like "make systemd" or something, that means no "replace" with reltool but otherwise might be not a bad idea.

Good point.
AFAIK, the packaging script for SmartOS (Solaris alike) has depended on **make release** and just copied the generated directories recursively like here https://github.com/leo-project/leofs/blob/master/pkg/leo_manager/Makefile#L22 so that it would be encouraged to move it to another make target to avoid bothering non-linux users as you suggested.

> There are few service files so I created new directory so they're together (some related files might be placed there in the future as well - e.g. limits configuration or helper scripts). Also, since systemctl supports nice output and autocompletion, shortening service names is discouraged, also _ is recommended to be replaced with -, so I took liberty in naming services for "leo_manager_0" and "leo_manager_1" as "leofs-manager-master" and "leofs-manager-slave".

Got it. looks reasonable.

**Comment by vstax** [comment]
@mocchira
Thank you for advice. It turned out that picking service files directly from rpm/deb install script doesn't have the downsides I thought it has. I'm still able to expand macro with reltool, by putting result into new directory which isn't copied into "package".

Package changes in new version:
1. For RPM package, `/usr/local/bin/leofs-adm` is real file again (no need for alternative with just one package version)
2. There is a new symlink `/usr/local/leofs/current -> /usr/local/leofs/<version>` to allow fixed paths.  It's a real symlink from package, no need for alternatives when only single version can be installed.
3. On systemd-based systems, epmd is **always** managed through systemd (using socket type of service), even when launching nodes through current scripts. This is required so that nodes never launch epmd by themselves when started, which allows user to do transition to using systemd services without a hassle at any time. This doesn't create any problems no matter how nodes are launched, but does create some issues during first upgrade to new version.

To ensure that epmd is always managed through systemd, package installs preset that enables `leofs-epmd.socket` by default (e.g. on next boot) and actually starts it from package script. Since it forces systemd to listen on the socket, it won't work if older, non-systemd managed instance of epmd is already running. So I'm thinking of suggesting upgrade procedure from version < 1.4.0 (tentatively) to 1.4.0 or higher as follows for systemd-based systems (EL7 and Ubuntu 16.04):
1. Stop currently running node. Backup current config files if needed.
2. Execute `ps aux | grep epmd | grep -v grep` and make sure no instances of epmd are running; if they are, kill them (e.g. `pkill epmd`). Make sure that LeoFS nodes on current system are stopped before killing epmd, though.
3. For EL7, execute `rpm -q leofs` and make sure that no more than single version of leofs package is installed.
4. Upgrade to new package version (`dpkg -i` for Debian, `rpm -U` for EL). After that, follow the usual procedure for LeoFS upgrade between versions.

If an error was made and epmd wasn't stopped before installation of the new package, an error similar to this will appear during package upgrade:
```
Job for leofs-epmd.socket failed. See "systemctl status leofs-epmd.socket" and "journalctl -xe" for details.
warning: %posttrans(leofs-1.4.0-1.x86_64) scriptlet failed, exit status 1
```
In that case, users will have to do the following:
1. After making sure that LeoFS nodes aren't running on this system, kill running epmd (e.g. `pkill epmd`)
2. Execute `systemctl start leofs-epmd.socket` - there should be no output if everything is done correctly

Nothing is changed for non-systemd users on Ubuntu 14. Users of EL6 need to make sure they only got single version installed and upgrade using `rpm -U`.

After 1.4.0 is installed and running, all next upgrades don't require any special steps.

Reasons for such upgrade procedure is that safely dealing with running epmd in automated way isn't really possible: if user forgot to stop node before upgrade, killing epmd would completely mess the things up. It would also conflict with cases like when someone wants to run different Erlang software on the same system which requires epmd as well and wants some other global instance of epmd - with the way it's currently done it's possible to mask `leofs-epmd.service` and override dependencies that require it everything will be compatible with future upgrades (this can be mentioned in documentation).


Removing old version and installing new one instead of upgrading is supported as well, of course. Installation and upgrade from older version follow slightly different rpm scripts - "%post" for initial installation and trigger for upgrades from older version; upgrades from package with systemd support to another package with systemd support don't require any scripts. %posttrans trigger always executes `systemctl start leofs-epmd.socket`, this is required for first time installation of first upgrade to systemd-supporting package, for other cases it's just a safeguard to get error right at that point if anything is wrong; if `leofs-epmd` is already running, it won't do anything at all.

`TimeoutStopSec=infinity` in service file doesn't work (it behaves the same as 0 timeout and kills process right away, not sure why), I set it to 30 minutes for now. On one hand, sometimes node can take a long time to shutdown under load, on the other hand there needs to be some limit in case of system reboot or something. 30 minutes is probably too much but I'm not sure what better values would be, at least not now.

The current version is available at
https://github.com/vstax/leofs/tree/systemd-exp-4 and https://github.com/vstax/leofs_package/tree/systemd (there are some hacks in like pulling out from my fork and "1.4.0" version hardcoded in trigger).
So far, I tested upgrades from all versions between 1.3.0 and 1.3.6 to this one - seems to work fine.
I haven't tested running full cluster using these services yet (need to test on individual systems and all-in-one). Also I haven't tested building and running on EL6 - current spec is written so there should be no changes except for the one forcing users to upgrade instead of installing versions side-by-side, but there might be mistakes. After making sure these things work, I'll start working on deb package.

I think after all this is done and tested I'll close this ticket, creating new ones for upstart support and for supporting proper services of "notify" type, since these need more work and can be done later separately.

**Comment by mocchira** [comment]
@vstax Thanks for your report.
Everything looks fine to me so please go ahead with the direction you've planed.

> TimeoutStopSec=infinity in service file doesn't work (it behaves the same as 0 timeout and kills process right away, not sure why), I set it to 30 minutes for now. On one hand, sometimes node can take a long time to shutdown under load, on the other hand there needs to be some limit in case of system reboot or something. 30 minutes is probably too much but I'm not sure what better values would be, at least not now.

I couldn't find the reason why infinity doesn't work however considering how it should behave with infinity when the system goes down remind me of the promise of infinity might be broken as it can cause the system to hang up in the middle of its termination if it keep its promise of infinity. so it maybe documentation problem in systemd. (I'm not sure)
anyway, 30 minutes looks reasonable for now.

> I think after all this is done and tested I'll close this ticket, creating new ones for upstart support and for supporting proper services of "notify" type, since these need more work and can be done later separately.

Got it.

**Comment by mocchira** [comment]
@vstax one thing I noticed **rel/service/leofs-epmd.service** generated by reltool should be included in .gitignore?

**Comment by vstax** [comment]
@mocchira The whole directory is in .gitignore (https://github.com/vstax/leofs/commit/a855030e4baa9ab0b8c803c828110adb94a48f08#diff-a084b794bc0759e7a6b77810e01874f2R4). It's a bit unobvious but `{copy,     "../service/leofs-epmd.service.pre", "../service/leofs-epmd.service"}` doesn't create copy in the same directory as these path arguments are relative to different directories, so the first directory here is `rel/service` but second will be `rel/leo_manager/service`. Which works just fine since I wasn't happy with creating generated file at the same place as the source.

Anyhow I fixed various bugs; most unobvious change is probably `After=leofs-epmd.service` added to service files, they don't require leofs-epmd.service to launch as they depend on leofs-epmd.socket instead, but this line tweaks shutdown order, so that leofs-epmd.service is stopped only after the real node has finished stopping, which is required for correct shutdown. Right now everything works as expected both on EL7 and EL6. I was powering off/on and rebooting nodes in various combination and it always worked fine in the end, no unexpected errors during shutdown or startup.

Starting VMs with whole cluster at once works fine (at least as long as there is no race between managers) because even when gateway and storage nodes don't see manager and shut down, they are restarted 5 seconds afterwards. Currently timeouts are: 5 second wait between restarts (no matter if they are caused by crash of already running node, or node that didn't finish starting up - it would require "notify" type of service to differentiate between these situations) and 30 minutes for shutdown.

I was thinking that while 30 minutes is OK for storage nodes, managers and gateway probably don't need that and a smaller one like 5 minutes would be better? (if any process are still present after that time, systemd kills them). Or there are some known conditions when gateways or managers need more time to shut down? I think it's possible for manager to get stuck on some leofs-adm operations but not sure it affects shutdown path.

It might be better to increase 5 seconds between restarts to something else like 10-30 seconds by default as well. Hard to say which is better, shorter time between restart for crash on working system or less logs when leo_storage is trying to start up and then exists in absence of manager. Given that it's a known case and all.

**Comment by mocchira** [comment]
@vstax 

> The whole directory is in .gitignore (vstax/leofs@a855030#diff-a084b794bc0759e7a6b77810e01874f2R4). It's a bit unobvious but {copy, "../service/leofs-epmd.service.pre", "../service/leofs-epmd.service"} doesn't create copy in the same directory as these path arguments are relative to different directories, so the first directory here is rel/service but second will be rel/leo_manager/service. Which works just fine since I wasn't happy with creating generated file at the same place as the source.

Got it. thanks for your explanation.

>Anyhow I fixed various bugs; most unobvious change is probably After=leofs-epmd.service added to service files, they don't require leofs-epmd.service to launch as they depend on leofs-epmd.socket instead, but this line tweaks shutdown order, so that leofs-epmd.service is stopped only after the real node has finished stopping, which is required for correct shutdown. Right now everything works as expected both on EL7 and EL6. I was powering off/on and rebooting nodes in various combination and it always worked fine in the end, no unexpected errors during shutdown or startup.

>Starting VMs with whole cluster at once works fine (at least as long as there is no race between managers) because even when gateway and storage nodes don't see manager and shut down, they are restarted 5 seconds afterwards. Currently timeouts are: 5 second wait between restarts (no matter if they are caused by crash of already running node, or node that didn't finish starting up - it would require "notify" type of service to differentiate between these situations) and 30 minutes for shutdown.

Got it.

> I was thinking that while 30 minutes is OK for storage nodes, managers and gateway probably don't need that and a smaller one like 5 minutes would be better? (if any process are still present after that time, systemd kills them). Or there are some known conditions when gateways or managers need more time to shut down? I think it's possible for manager to get stuck on some leofs-adm operations but not sure it affects shutdown path.

Sounds good to choose a smaller one.
A long period like 30 minutes should never be required on manager/gateway(s) as there is much less I/O bound tasks compared to storage(s). 

> It might be better to increase 5 seconds between restarts to something else like 10-30 seconds by default as well. Hard to say which is better, shorter time between restart for crash on working system or less logs when leo_storage is trying to start up and then exits in the absence of manager. Given that it's a known case and all, might want to smooth things a bit by increasing sleep time between restarts.

Sounds good to increase 5 second.

**Comment by vstax** [comment]
@mocchira 

I tweaked timeouts, and added some dependencies between nodes. They are ignored for single-node-per-server configuration and only activate for all-in-one configuration, though (once again) they won't be 100% reliable at start without "notify" service type, but at least they improve shutdown order (gateway/storage shutdown before master/slave, slave shutdowns before master).

There is a slight problem with building deb package - I couldn't find a way similar to macros from rpm version which allow to keep the same spec for both systemd and non-systemd versions. For deb, it seems kind of impossible to reliably know if we're supposed to build version for systemd or non-systemd OS: it is possible to check if current system is running systemd e.g. with `[ -d /run/systemd/system ]`, but it won't work in container/chroot and similar cases, so it's better to specify it explicitly. So I think now there will be either two different scripts for systemd and non-systemd version, or accepted parameter which tweaks the inside a bit. Of course, parameter (e.g. `./make_deb.sh 1.4.0 true`) is nicer, but additions required to build systemd package - to scripts and makefile and some others - are quite extensive. You can look at them here: https://github.com/vstax/leofs_package/commit/c03990c63249192f53cba894198b7dcd87f8806b (some changes here like more dependencies are actually general ones, though, but most is systemd-related). For now I'm just not sure how it would be possible to make two-in-one script with this amount of changes (and keep it nice enough) with just bash variables, would be nice to have a proper macro processor here, I think. Need to think about this a bit more.

Btw, it would be best to check if epmd is build with systemd support for building (e.g.  by calling "epmd --help 2>&1 |grep systemd" or "ldd /usr/lib/erlang/erts-7.3/bin/epmd|grep systemd"), because building package for systemd-based system with Erlang runtime where epmd wasn't compiled with systemd support will result in broken package which won't work. Might be a good idea to add the check to script (in case of deb package) and rpm spec itself (since variable which picks whether we're build for systemd or not is defined in spec itself).


About Ubuntu 16.04 packaging scripts. It's much simpler in RHEL which completely moved from upstart in 6 to systemd in 7 and everything was changed so there are just two possible states (either it's EL6, then system is running upstart or it's EL7, then systemd is fully supported and the system is running it) and simple rpm macros which help you to manage systemd-related stuff, plus systemctl can be used directly on EL7. However, Debian and Ubuntu had a gradual migration over to systemd, and there was a period (e.g. Debian 8) when systemd might be present or there might be another init system in place of it, and the fact that it's installed doesn't mean that it's used. Because of that dh-systemd (separate package to debhelper 9 used in Ubuntu 16, but merged into main debhelper for versions >= 10) offers some elaborate logic to write systemd-compatible scripts, which together with wrapper called `deb-systemd-helper` is designed so it's possible to write code which will work both on systemd and non-systemd system (there is also initscripts generator from systemd services and various other stuff).

The price for that is that these scripts and wrappers use external state to track which services are supposed to be running between package installs/uninstalls and upgrades instead of regular systemd mechanism. So packages are supposed to use these instead of e.g. systemd presets. These install scripts aren't fully compatible with .socket units as well, I think (at very least, not everything works in dh-systemd from Ubuntu 16.04). Since leofs packages supply bunch of services, socket and preset and require certain logic, I can't find a simple way to use dh-systemd helpers and make everything work the way it's needed to be (not to mention that presets aren't quite functional because Debian and Ubuntu ship with "enable *" preset by default and shipping disabling preset in leofs package creates harmless but annoying warnings due to bug in systemd). Not saying that using helpers for this case is impossible but it's non-trivial at least, I just couldn't make some stuff to work (e.g. handling the case when user has masked leofs-epmd.socket to use some other epmd installation instead). So I decided to supply all the logic I need directly in scripts, without dh-systemd helpers. Maybe someone who has better understanding how to make it all compatible with these helpers later can improve on that, but for now I had to settle for code written directly around "systemctl" without helpers. It's not complicated or anything, the main downside is that it's not compatible with some rare cases like someone picking systemd-supporting package and using it on Debian or Ubuntu where systemd was manually disabled or replaced by some other init system - packages using wrappers might be compatible with that but not leofs package, at least not now. Another side-effect is that removing package, then installing it again would reset state of services (e.g. if someone installed package, then enabled leofs-manager-master, then removed package, then installed the package again - the service will be disabled, just like for rpm package) - I think it would behave differently when using helpers which save "enabled" state even for removed packages, but this is probably not a big deal.

That said, in general systemd services work fine on Ubuntu 16.06 and there seems to be no issues both for first installation and upgrade (as long as epmd is killed before upgrade, that is).

**Comment by vstax** [comment]
TODO:

- [ ] Merge non-systemd and systemd versions of "make_deb.sh" in a nice way 
- [ ] Check for epmd with systemd support before building systemd-supported rpm and deb
- [ ] Write documentation for users about upgrade
- [ ] Write documentation for users about using systemd services and changes (do not mix launching with normal scripts and through service, contents of erlang.log moved to "journalctl -u .." and so on)
- [ ] Notify users in ML about changes, link to documentation
- [ ] Merge patches from main repo (actually they are 100% compatible and can be merged at any time, but I'm tweaking the service files themselves from time to time still, I'll make PR when I'm sure they are in final form)
- [ ] Merge packaging changes, build packages with systemd support, replace "1.4.0" with different version if the merge happens before some other 1.3.x release
- [ ] Final testing (I'll need some help here)

**Comment by mocchira** [comment]
@vstax Thanks for your detailed explanation. everything (your decisions how to implement) looks fine to me so please go ahead with your plan.

**Comment by vstax** [comment]
@mocchira 

I posted PR with squashed commits to main repo (#829), it's supposed to be fully compatible and not to break anything and can be merged if everything is OK.

I wrote documentation, it's available at https://github.com/vstax/leofs/tree/systemd-docs
(just like leofs_package change, it will need 1.4.0 replaced to actual version released with systemd support). Documentation can be merged after package changes are merged, I think.

Regarding packages, currently deb and rpm use different package descriptions, rpm has
```
%description
LeoFS is a highly available, distributed, eventually consistent object/blob store. Organizations can use LeoFS to store lots of data efficiently, safely, and inexpensive.
```
and deb has
```
Description: LeoFS is the Web shaped object storage system.
 LeoFS is the Web shaped object storage system and S3 compatible storage.
```
I would be nicer to have them the same... Also "Web shaped" sounds a bit strange, I think.. Should I change / merge them to be the same? For e.g.
```
LeoFS is a highly available, distributed, eventually consistent S3 compatible storage. Organizations can use LeoFS to store lots of data efficiently, safely, and inexpensive.
```

**Comment by mocchira** [comment]
@vstax 

> I posted PR with squashed commits to main repo (#829), it's supposed to be fully compatible and not to break anything and can be merged if everything is OK.

LGTM. I will merge that later.

> I wrote documentation, it's available at https://github.com/vstax/leofs/tree/systemd-docs
(just like leofs_package change, it will need 1.4.0 replaced to actual version released with systemd support). Documentation can be merged after package changes are merged, I think.

Got it.

> I would be nicer to have them the same... Also "Web shaped" sounds a bit strange, I think.. Should I change / merge them to be the same? For e.g.

Yes please change to what you suggested.

**Comment by vstax** [comment]
@mocchira 
Thank you for review. I created PR for package changes https://github.com/leo-project/leofs_package/pull/7
The current devel branch in repo should be tagged as 1.2.0 (similar to how it was done during package changes for 1.3.3) before merge. I changed check_version.sh to track this so there will be a warning if the new package is used to build 1.3.7 or older.

There are two places (one in `leofs.spec` and another in `make_deb.sh`) that refer to "1.4.0", they will need to be changed to 1.3.8 if decision about that release is made. Or, alternatively, they can be changed to "1.3.8" even now, it will work correctly even if next release will be 1.4.0, just will be slightly confusing maybe.

**Comment by mocchira** [comment]
@vstax 

> The current devel branch in repo should be tagged as 1.2.0 (similar to how it was done during package changes for 1.3.3) before merge. I changed check_version.sh to track this so there will be a warning if the new package is used to build 1.3.7 or older.

OK.

> EDIT: actually I thought about it and just changed it to 1.3.8, so there will be no more need to change it no matter what next version will be. Fixed a small issue with deb scripts as well (it was incorrect not to have a shell in new scripts, though it worked well in practice).

Got it.

**Comment by mocchira** [comment]
Done tagging the current develop as 1.2.0.

**Comment by vstax** [comment]
NB: Add LimitNOFILE=65535 to default unit files for leofs-gateway and leofs-storage, no need to make half of users override default limit of 1024.

**Comment by vstax** [comment]
@mocchira 
[I've prepared text for ML notification](https://gist.github.com/vstax/45c50901a145bbd8bd707d891d5e515a)
I plan to replace current links to documentation on official site after #862 is merged and latest version of documentation is rebuilt / released and send the message after that. Please share your thoughts, if everything is OK or something needs changing.

**Comment by mocchira** [comment]
@vstax thanks! LGTM. I dropped a comment pointing out a few minor fixes on https://gist.github.com/vstax/45c50901a145bbd8bd707d891d5e515a#gistcomment-2220017

**Comment by vstax** [comment]
@mocchira Thank you, I did these and few other fixes.

**Comment by mocchira** [comment]
@vstax as you may know, the doc has been published on the official site so please go ahead.

**Comment by vstax** [comment]
@mocchira Right, I've sent the message.
Besides some extra testing, this issue as well as #791 and #792 are done.

**Comment by vstax** [comment]
@mocchira This, as well as two other issues probably can be re-tagged as 1.3.8 and closed after you did extra testing that might be needed; all features are working fine on my systems and I don't see any issues (we use only RPM versions of packages, though). I just did some checks and I think that everything will work correctly when creating 1.3.8-rc1 package and using it for testing upgrades from previous versions and everything else, the upgrade triggers conditions should work exactly as they do for the final 1.3.8.

Of course, 1.3.8-rc1 can't be upgraded to 1.3.8 without forcing after that because both rpm/dpkg consider it to be higher version than 1.3.8 but that's issue of current naming scheme and outside the scope of package script and triggers (not that anything here should break when 1.3.8-rc1 is "downgraded" - from package perspective - to 1.3.8 anyway).

**Comment by mocchira** [comment]
I finished the final testing on FreeBSD/SmartOS with each latest version. as there is no complains about this change, I will close this issue. 

Feel free to reopen if you find something to be done/fixed before 1.3.8 get released.
